Lib/annotationlib.py

```diff
@@ -230,18 +218,30 @@ def __forward_code__(self):
     def __eq__(self, other):
         if not isinstance(other, ForwardRef):
             return NotImplemented
-        if self.__forward_evaluated__ and other.__forward_evaluated__:
-            return (
-                self.__forward_arg__ == other.__forward_arg__
-                and self.__forward_value__ == other.__forward_value__
-            )
         return (
             self.__forward_arg__ == other.__forward_arg__
             and self.__forward_module__ == other.__forward_module__
+            and self.__forward_is_class__ == other.__forward_is_class__
+            and self.__code__ == other.__code__
+            and self.__ast_node__ == other.__ast_node__
+            # Use "is" here because we use id() for this in __hash__
+            # because dictionaries are not hashable.
+            and self.__globals__ is other.__globals__
+            and self.__cell__ == other.__cell__
+            and self.__owner__ == other.__owner__
         )
```

Comment:
you could possibly micro-optimise this slightly by moving some of the cheaper comparisons up (e.g. the `self.__globals__ is other.__globals__` comparison should be very cheap since it's just an identity check).

It took me a while to wrap my head around the `self.__globals__ is other.__globals__` comparison but I _think_ it's correct to do it that way...!

Reply:
Reordered them a bit. I think the performance effects are quite hard to predict though because it matters not just how expensive the check is, but also how likely it is to short-circuit the comparison. For example, maybe since we already checked the module, the globals check actually never makes a difference, because if the module doesn't match the globals also won't match.

If somebody wants to micro-optimize this I won't stop them but I'd rather just land this; I doubt it will make a huge difference either way.